src/de/mod.rs

```diff
@@ -1672,10 +1673,185 @@
+//!     ($impl_type: ty, $tag: literal, $(($variant_tag: literal, $typ: tt, $variant: expr, $enum_type: ident)),+) => {
+//!         impl<'de> ::serde::Deserialize<'de> for $impl_type {
+//!             fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+//!             where
+//!                 D: ::serde::Deserializer<'de>,
+//!             {
+//!                 use ::serde::de::Error;
+//!                 // The Visitor struct is normally used for state, but none is needed
+//!                 struct TheVisitor;
+//!
+//!                 // The main logic of the deserializing happens in the Visitor trait
+//!                 impl<'de> ::serde::de::Visitor<'de> for TheVisitor {
+//!
+//!                     // The type that is being deserialized
+//!                     type Value = $impl_type;
+//!
+//!                     // try to give a better error message when this is used wrong
+//!                     fn expecting(&self, formatter: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+//!                         formatter.write_str(&format!("expecting map with tag in {}", $tag))
+//!                     }
+//!
+//!                     // the xml data is provided as an opaque map, that map is parsed into the type
+//!                     fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+//!                     where
+//!                         A: ::serde::de::MapAccess<'de>,
+//!                     {
+//!                         // the first entry in the map is the attribute
+//!                         let entry: Option<(String, String)> = map.next_entry()?;
+//!                         // here the assumption is made that only one attribute exists and it's the tag
+//!                         // if there are more attributes those would need to be parsed as well
+//!                         let tag = match entry {
+//!                             // return an error if the no attributes are found, and indicate that the @tag attribute is missing
+//!                             None => Err(A::Error::missing_field($tag)),
+//!                             // check if the attribute is the tag
+//!                             Some((attribute, value)) => {
+//!                                 if attribute == $tag {
+//!                                     // return the value of the tag
+//!                                     Ok(value)
+//!                                 } else {
+//!                                     // the attribute is not @tag, return an error indicating that there is an unexpected attribute
+//!                                     Err(A::Error::unknown_field(&attribute, &[$tag]))
+//!                                 }
+//!                             }
+//!                         }?;
+//!
+//!                         // get the next key in the map, this should be the variant of the enum
+//!                         let key: Option<String> = map.next_key()?;
```

Comment:
Actually, I have a feeling, that next code can be simplified and get more generic if using [`MapAccessDeserializer`](https://docs.rs/serde/latest/serde/de/value/struct.MapAccessDeserializer.html)

Something like:
```rust
let de = MapAccessDeserializer::new(map);
match tag {
  $($variant_tag => $typ::deserialize(de).map($variant)),+
}
```

Although, the current variant is ok for demonstration purposes.

Reply:
This only works for newtype variants, I've left it out for now